### PR TITLE
#156: Connects Exam front- and back-end

### PR DESCRIPTION
**Pull request by sandroroux**
Closes https://github.com/catalpainternational/asteroid/issues/156.

Runs against https://github.com/catalpainternational/canoe-backend/pull/49 and https://github.com/catalpainternational/asteroid/pull/147.

### CODE REVIEW

File: src/riot/WagtailPages/CoursePage.riot.html [hunk 1]
```diff
@@ -26,84 +26,63 @@ <h5 class="section-title" translate>Your lessons</h5>
...
+                    isComingSoon="{false}"
+                    showFooter="{true}"
+                    isExam="{true}"
+                ></Card>
+            </template>
         </div>
     </div>
     <CourseExam
-        if="{state.hash === examModule}"
+        if="{state.hash === EXAM_MODULE && state.examCards.length > 0}"
```
Comment (katevoss) [inline src/riot/WagtailPages/CoursePage.riot.html:45]: Why did you add the conditional `state.examCards.length > 0`?

You might want to add this conditional to the "Take exam" button on `<LessonComplete>` when a user is directed to take the Course Exam. That way the button only appears if there is an exam.

File: src/riot/WagtailPages/CoursePage.riot.html [hunk 2]
```diff
@@ -26,84 +26,63 @@ <h5 class="section-title" translate>Your lessons</h5>
...
-                    answers: [
-                        { correct: false, text: "ba dum dum dum" },
-                        { correct: true, text: "dee ba da" },
-                        { correct: false, text: "ba dum dum dum" },
-                        { correct: true, text: "dee ba da" },
-                    ],
-                    question: "What does Karen sing?",
-                },
-            ],
+            EXAM_MODULE: "exam",
```
[the remark this answers is not in the document]
Comment (PeteCoward) [reply src/riot/WagtailPages/CoursePage.riot.html:71]: @sandroroux loves a nitpick :)

[the remark this answers is not in the document]
Comment (sandroroux) [reply src/riot/WagtailPages/CoursePage.riot.html:71]: @katevoss, that's a good point. I think I messed around with a conditional, which is why I changed the variable name.

As we discussed, the codebase lacks an agreed-upon style guide. In general, I try to write variables with unchanging values in uppercase.

Uppercase constants are [typical](https://stackoverflow.com/questions/34732630/all-caps-constants-in-javascript-and-requireds-and-imports) and, I think, add meaning. I like [this guy's take](https://www.freecodecamp.org/news/when-to-capitalize-your-javascript-constants-4fabc0a4a4c4/) too because he suggests that uppercase constants are optional, like all practices.

I'd rather the team agree on an approach. I'm for consistency between `CourseExam` and `LessonPage`, and will revert if you think the other approach is better.

File: src/riot/WagtailPages/CoursePage.riot.html (hunk 1 above)
Comment (sandroroux) [reply src/riot/WagtailPages/CoursePage.riot.html:45]: Someone could navigate to the URL when there are no exam cards, and errors would ensue. I should rethink this. The problem probably has a better solution.

Thanks for the heads-up on `LessonComplete`. I'll take a look.

File: src/riot/WagtailPages/CoursePage.riot.html [hunk 3]
```diff
@@ -26,84 +26,63 @@ <h5 class="section-title" translate>Your lessons</h5>
...
-                isCompleted="{false}"
-                isComingSoon="{false}"
-                showFooter="{true}"
-                isExam="{true}"
-            ></Card>
+            <template if="{state.examCards.length > 0}">
+                <h5 class="section-title" translate>Course exam</h5>
+                <Card
+                    title="{`${props.wagtailPage.title}'s Course Exam`}"
+                    description="{'this is the course exam description'}"
```
Comment (katevoss) [inline src/riot/WagtailPages/CoursePage.riot.html:33]:
```suggestion
```
There isn't (currently) a description of the course exam in the backend so let's remove this altogether.

File: src/riot/WagtailPages/CoursePage.riot.html [hunk 4]
```diff
@@ -26,84 +26,63 @@ <h5 class="section-title" translate>Your lessons</h5>
...
-                link="{props.wagtailPage.id}/exam/1"
-                isCompleted="{false}"
-                isComingSoon="{false}"
-                showFooter="{true}"
-                isExam="{true}"
-            ></Card>
+            <template if="{state.examCards.length > 0}">
+                <h5 class="section-title" translate>Course exam</h5>
+                <Card
+                    title="{`${props.wagtailPage.title}'s Course Exam`}"
```
Comment (katevoss) [inline src/riot/WagtailPages/CoursePage.riot.html:32]: Adding the "[CourseTitle]'s Course Exam" feels a little redundant as there is also a section title "Course Exam" right above the card. 
image

I would recommend removing it from the title, but I'm not too bothered so it's not a required change. Do you have any thoughts?

File: src/riot/WagtailPages/CoursePage.riot.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (katevoss) [reply src/riot/WagtailPages/CoursePage.riot.html:71]: Consistency is my number 1 priority and second to that I think writing variables that represent constants as UPPERCASE is a good approach. If you could make the relevant changes in `LessonPage` too then I'm on board!
